dsgrid/query/models.py

```diff
@@ -0,0 +1,396 @@
+import abc
+import enum
+from pathlib import Path
+from typing import Any, List, Optional, Set, Union
+
+import pyspark.sql.functions as F
+from pydantic import Field, root_validator, validator
+from semver import VersionInfo
+
+from dsgrid.data_models import DSGBaseModel
+
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.dimension.dimension_filters import make_dimension_filter, DimensionFilterBaseModel
+from dsgrid.utils.files import compute_hash, load_data
+
+
+class FilteredDatasetModel(DSGBaseModel):
+
+    dataset_id: str = Field(description="Dataset ID")
+    filters: List[Any] = Field(
+        description="Dimension filters to apply to the dataset'",
+    )
+
+
+# TODO: setting alias on a groupBy column doesn't work on a per-dataset aggregation
+
+
+class ColumnModel(DSGBaseModel):
+    """Defines one column in a SQL aggregation statement."""
+
+    dimension_query_name: str
+    function: Optional[Any] = Field(
+        description="Function or name of function in pyspark.sql.functions."
+    )
+    alias: Optional[str] = Field(description="Name of the resulting column.")
+
+    @validator("function")
+    def handle_function(cls, function_name):
+        if function_name is None:
+            return function_name
+        if not isinstance(function_name, str):
+            return function_name
+
+        func = getattr(F, function_name, None)
+        if func is None:
+            raise ValueError(f"function={function_name} is not defined in pyspark.sql.functions")
+        return func
+
+    @validator("alias")
+    def handle_alias(cls, alias, values):
+        if alias is not None:
+            return alias
+
+        func = values.get("function")
+        if func is not None:
+            name = values["dimension_query_name"]
+            return f"{func.__name__}__{name}"
+
+        return alias
+
+    def dict(self, *args, **kwargs):
+        data = super().dict(*args, **kwargs)
+        if data["function"] is not None:
+            data["function"] = data["function"].__name__
+        return data
+
+    def get_column_name(self):
+        if self.alias is not None:
+            return self.alias
+        if self.function is None:
+            return self.dimension_query_name
+        return f"{self.function.__name__}__{self.dimension_query_name})"
+
+
+class DimensionQueryNamesModel(DSGBaseModel):
+    """Defines columns to include in a SQL aggregation statement. If a value is empty, that
```

Comment:
Suggested addition: This defines the list of dimensions the value columns should be aggregated to.